pilot/pkg/config/memory/controller.go

```diff
@@ -36,6 +38,7 @@ func NewController(cs model.ConfigStore) model.ConfigStoreCache {
 	out := &controller{
 		configStore: cs,
 		monitor:     NewMonitor(cs),
+		mutex:       sync.RWMutex{},
```

Comment:
can you explain what the mutex is needed for?

Reply:
Without this in tests where create of service entries + read via the new edsUpdate for added service entries (that in turn triggers maybeRefreshUpdate because it is a full push) is causing a data race - specially in TestAgent (in ads_test) and TestServiceDiscoveryServiceUpdate (servicediscovery_test)

Reply:
Do you have any other thoughts of avoiding it?

Reply:
Here is an example race condition -  this happens when you create config and immediately call some method like GetProxyInstances which triggers the refresh indexes (which will be true initially)
```
runtime.mapassign_faststr()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/map_faststr.go:202 +0x0
  istio.io/istio/pilot/pkg/config/memory.(*store).Create()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/config/memory/config.go:173 +0x892
  istio.io/istio/pilot/pkg/config/memory.(*controller).Create()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/config/memory/controller.go:81 +0xd3
  istio.io/istio/pilot/pkg/model.(*istioConfigStore).Create()
      <autogenerated>:1 +0xe3
  istio.io/istio/pilot/pkg/serviceregistry/serviceentry.createConfigs()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/serviceregistry/serviceentry/servicediscovery_test.go:43 +0x154
  istio.io/istio/pilot/pkg/serviceregistry/serviceentry.TestServiceDiscoveryGetProxyServiceInstances()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/serviceregistry/serviceentry/servicediscovery_test.go:751 +0x122
  testing.tRunner()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:991 +0x1eb

Previous read at 0x00c000f1a0f0 by goroutine 75:
  runtime.mapiternext()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/map.go:846 +0x0
  istio.io/istio/pilot/pkg/config/memory.(*store).List()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/config/memory/config.go:117 +0x2de
  istio.io/istio/pilot/pkg/config/memory.(*controller).List()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/config/memory/controller.go:116 +0xe1
  istio.io/istio/pilot/pkg/model.(*istioConfigStore).ServiceEntries()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/model/config.go:383 +0x100
  istio.io/istio/pilot/pkg/serviceregistry/serviceentry.(*ServiceEntryStore).maybeRefreshIndexes()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:497 +0x1de
  istio.io/istio/pilot/pkg/serviceregistry/serviceentry.(*ServiceEntryStore).edsUpdate()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:438 +0x202
  istio.io/istio/pilot/pkg/serviceregistry/serviceentry.(*ServiceEntryStore).serviceEntryHandler()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:246 +0x5d4
  istio.io/istio/pilot/pkg/serviceregistry/serviceentry.(*ServiceEntryStore).serviceEntryHandler-fm()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:150 +0x135
  istio.io/istio/pilot/pkg/config/memory.(*configstoreMonitor).applyHandlers()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/config/memory/monitor.go:103 +0x1c6
  istio.io/istio/pilot/pkg/config/memory.(*configstoreMonitor).processConfigEvent()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/config/memory/monitor.go:94 +0x179
  istio.io/istio/pilot/pkg/config/memory.(*configstoreMonitor).Run()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/config/memory/monitor.go:83 +0x220
  istio.io/istio/pilot/pkg/config/memory.(*controller).Run()
      /Users/rama.rao/git/oss/istio/src/istio.io/istio/pilot/pkg/config/memory/controller.go:69 +0x5a
```

Reply:
As it shows, it is a bug of memory store. We should fix the memory config store to support r/w concurrently

Reply:
Ok, I moved the mutex to memory store instead of controller